### PR TITLE
feat[venom]: optimize `get_basic_block()`

### PR DESCRIPTION
**Pull request by charles-cooper**
### Commit message

```
`get_basic_block()` is a hotspot in venom (up to 35% of total
compilation time!). this optimizes `get_basic_block()`, on a large
contract near the 24kb limit this reduces time spent in venom from 3s to
1s (total time from 6s to 4s).

note on the same contract, time spent in the IRnode optimizer pipeline
is 2s - so time in venom is now smaller than time in legacy optimizer(!)

notes:
- refactor to use dict for basic_blocks
- clean up basic blocks API 
    hide basic blocks behind `get_basic_blocks()` iterator and
    `num_basic_blocks`.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/venom/function.py [hunk 1]
```diff
@@ -40,30 +40,35 @@ def __init__(self, name: IRLabel, ctx: "IRContext" = None) -> None:  # type: ign
...
+
+        self.basic_blocks.remove(bb)
+
+        self._bb_index.pop(bb.label, None)
+
+    def _get_basicblock(self, label: str):
+        bb = self._bb_index.get(label)
+        if bb is None:
+            # search for it the slow way, indexing as we go along
+            for bb in self.basic_blocks:
```
Comment (harkal) [inline vyper/venom/function.py:63]: CQ: Maybe not shadow the outer bb?

File: vyper/venom/ir_node_to_venom.py [hunk 2]
```diff
@@ -137,8 +137,7 @@ def _append_jmp(fn: IRFunction, label: IRLabel) -> None:
 
 def _new_block(fn: IRFunction) -> IRBasicBlock:
```
Comment (harkal) [inline vyper/venom/ir_node_to_venom.py:138]: Update the return signature now that the function does not return anything, or return the new block anyway?

Comment (charles-cooper) [reply vyper/venom/ir_node_to_venom.py:138]: i think didn't get a chance to run the linter yet!

File: vyper/venom/function.py [hunk 3]
```diff
@@ -142,7 +124,7 @@ def _compute_reachability(self) -> None:
         """
         Compute reachability of basic blocks.
         """
-        for bb in self.basic_blocks:
+        for bb in self._basic_block_dict.values():
```
Comment (harkal) [inline vyper/venom/function.py:127]: Use `get_basic_blocks()` for uniformity?

Comment (charles-cooper) [reply vyper/venom/function.py]: https://github.com/vyperlang/vyper/pull/4002/commits/f8238c3670691907bc68c013bdbf0a189896661e

**Review by harkal — APPROVED** [review]

**Pull request merged**